doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,903 @@
+
+```c
+/* Returns milliseconds or -1 (infinity). */
+int64_t SSL_get_tick_timeout(SSL *ssl);
+```
+
+Finally, the tick function allows the QUIC state machine to be ticked without
+performing any specific operation:
+
+```c
+int SSL_tick(SSL *ssl);
+```
+
+Equivalent BIO methods will be defined for each of the above.
+
+#### Supporting Non-Pollable BIOs
+
+Where we are provided with a non-pollable BIO, we cannot provide the application
+with any primitive used for synchronisation and it is assumed that the
+application will handle its own network I/O, for example via a `BIO_dgram_pair`.
+
+When libssl calls `BIO_get_[rw]poll_descriptor` on the underlying BIO, the call
+fails, indicating that a non-pollable BIO is being used. Thus, if an application
+calls `SSL_get_[rw]poll_descriptor`, that call also fails.
+
+`SSL_get_tick_timeout` is still functional and must be handled by the
+application.
+
+There are various circumstances which need to be handled:
+
+  - The QUIC implementation wants to write data to the network but
+    is currently unable to (e.g. `BIO_dgram_pair` is full).
+
+    This is not hard as our internal TX record layer allows arbitrary buffering.
+    The only limit comes when QUIC flow control (which only applies to
+    application stream data) applies a limit; then we must return
+    `SSL_ERROR_WANT_WRITE`.
+
+  - The QUIC implementation wants to read data from the network
+    but is currently unable to (e.g. `BIO_dgram_pair` is empty).
+
+    Here we need to return `SSL_ERROR_WANT_READ`; we thereby support libssl's
+    classic nonblocking I/O interface.
+
+### Configuration of Blocking vs. Non-Blocking Mode
+
+Traditionally an SSL object has operated either in blocking mode or non-blocking
+mode without requiring explicit configuration; if a socket returns EWOULDBLOCK
+or similar, it is handled appropriately, and if a socket call blocks, there is
+no issue. Since the QUIC implementation is building on non-blocking I/O, this
+implicit configuration of non-blocking mode is not feasible.
+
+Note that Windows does not have an API for determining whether a socket is in
+blocking mode, so it is not possible to use the initial state of an underlying
+socket to determine if the application wants to use non-blocking I/O or not.
+Moreover this would undermine the BIO abstraction.
+
+As such, an explicit call is introduced to configure an SSL (QUIC) object into
+non-blocking mode:
+
+```c
+int SSL_set_blocking_mode(SSL *s, int blocking);
+int SSL_get_blocking_mode(SSL *s);
+```
+
+Applications desiring non-blocking operation will need to call this API to
+configure a new QUIC connection accordingly. Blocking mode is chosen as the
+default for parity with traditional Berkeley sockets APIs and to make things
+simpler for blocking applications, which are likely to be seeking a simpler
+solution. However, blocking mode cannot be supported with a non-pollable BIO,
+and thus blocking mode defaults to off when used with such a BIO.
```

Comment:
It could be supported. We can simulate blocking. Although we've never done this before (e.g. with a mem bio) so this is possibly not a requirement.

Reply:
Right, this did occur to me. We *could* add a blocking mode to BIO_dgram_pair if we really wanted to, where one thread can unblock it on another thread by writing to it. But it's not something we've ever supported before, so I we don't need to start supporting it now, certainly not for MVP.

Reply:
So perhaps rather than "cannot be supported", just "is not supported".